src/main/java/org/openjdk/jextract/impl/SourceFileBuilder.java

```diff
@@ -0,0 +1,124 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.jextract.impl;
+
+import javax.tools.JavaFileObject;
+import java.lang.constant.ClassDesc;
+import java.util.function.Function;
+
+public class SourceFileBuilder {
+
+    private static final boolean SHOW_GENERATING_CLASS = Boolean.getBoolean("jextract.showGeneratingClass");
+
+    private final String packageName;
+    private final String className;
+
+    // code buffer
+    private final StringBuilder sb = new StringBuilder();
+    // current line alignment (number of 4-spaces)
+    private int align;
+
+    private SourceFileBuilder(String packageName, String className) {
+        this.packageName = packageName;
+        this.className = className;
+    }
+
+    public String className() {
+        return className;
+    }
+
+    public static SourceFileBuilder newSourceFile(String packageName, String className) {
+        SourceFileBuilder sfb = new SourceFileBuilder(packageName, className);
+        sfb.emitPackagePrefix();
+        sfb.emitImportSection();
+        return sfb;
+    }
+
+    void emitPackagePrefix() {
+        assert packageName.indexOf('/') == -1 : "package name invalid: " + packageName;
+        append("// Generated by jextract");
+        if (SHOW_GENERATING_CLASS) {
+            append(" (via ");
+            append(getClass().getName());
+            append(")");
+        }
+        append("\n\n");
+        if (!packageName.isEmpty()) {
+            append("package ");
+            append(packageName);
+            append(";\n\n");
+        }
+    }
+
+    void emitImportSection() {
+        append("import java.lang.invoke.MethodHandle;\n");
+        append("import java.lang.invoke.VarHandle;\n");
+        append("import java.nio.ByteOrder;\n");
+        append("import java.lang.foreign.*;\n");
+        append("import static java.lang.foreign.ValueLayout.*;\n");
+        append("\n");
+    }
+
+
+    // Internal generation helpers (used by other builders)
+
+    void append(String s) {
+        sb.append(s);
+    }
+
+    void append(char c) {
+        sb.append(c);
+    }
+
+    void append(long l) {
+        sb.append(l);
+    }
+
+    void indent() {
+        for (int i = 0; i < align; i++) {
+            append("    ");
+        }
+    }
+
+    void incrAlign() {
+        align++;
+    }
+
+    void decrAlign() {
+        align--;
+    }
+
+    int align() {
+        return align;
+    }
+
+    public JavaFileObject toFile(Function<String, String> finisher) {
+        return Utils.fileFromString(packageName, className, finisher.apply(sb.toString()));
+    }
```

Comment:
Used by `TopLevelBuilder::toFiles` to replace the placeholder `#{SUPER}` super class name for the first header.